alf/algorithms/misc_algorithm.py

```diff
@@ -0,0 +1,239 @@
+                    agent_pose, [3, 3], axis=-1)
+                joint_pose_1, joint_pose_2, joint_vel_1, joint_vel_2 = tf.split(
+                    internal_states, [1, 1, 1, 1], axis=-1)
+                joint_1 = tf.concat([joint_pose_1, joint_vel_1], axis=-1)
+                joint_2 = tf.concat([joint_pose_2, joint_vel_2], axis=-1)
+
+                if self.n_objects == 1:
+                    obs_achieved_goal = task_specific_ob
+                    obs_excludes_goal = agent_pose_1
+                    if self.empowerment:
+                        obs_excludes_goal = action
+                elif self.n_objects == 0:
+                    obs_achieved_goal = joint_1
+                    obs_excludes_goal = joint_2
+
+                return (obs_excludes_goal, obs_achieved_goal)
+
+            elif (dimo == 24) and (self.n_objects == 2):
+                task_specific_ob_1, task_specific_ob_2, agent_pose, agent_vel, internal_states, action = tf.split(
+                    o, [3, 3, 6, 6, 4, 2], axis=-1)
+
+                agent_pose_1, agent_pose_2 = tf.split(
+                    agent_pose, [3, 3], axis=-1)
+
+                obs_achieved_goal_1 = task_specific_ob_1
+                obs_achieved_goal_2 = task_specific_ob_2
+                obs_excludes_goal = agent_pose_1
+
+                return (obs_excludes_goal, obs_achieved_goal_1,
+                        obs_achieved_goal_2)
+            else:
+                print(
+                    'Please specify the state of interests and the context state in the misc_alorighm.py first.'
+                )
+                exit()
+        else:
+            print(
+                'Please specify the state of interests and the context state in the misc_alorighm.py first.'
+            )
+            exit()
+
+    def mine(self, x_in, y_in):
+
+        # Mutual Information Neural Estimation
+        y_in_tran = tf.transpose(y_in, perm=[1, 0, 2])
+        y_shuffle_tran = tf.gather(
+            y_in_tran, tf.random.shuffle(tf.range(tf.shape(y_in_tran)[0])))
+        y_shuffle = tf.transpose(y_shuffle_tran, perm=[1, 0, 2])
+        x_conc = tf.concat([x_in, x_in], axis=-2)
+        y_conc = tf.concat([y_in, y_shuffle], axis=-2)
+
+        # propagate the forward pass
+        layerx = self._misc_layerx(inputs=x_conc)
+        layery = self._misc_layery(inputs=y_conc)
+        layer2 = tf.nn.relu(layerx + layery)
+        output = self._misc_layero(inputs=layer2)
+        output = tf.nn.tanh(output)
+
+        # split in T_xy and T_x_y predictions
+        N_samples = tf.shape(x_in)[-2]
+        T_xy = output[:, :N_samples, :]
+        T_x_y = output[:, N_samples:, :]
+
+        # compute the negative loss (maximise loss == minimise -loss)
+        mean_exp_T_x_y = tf.reduce_mean(tf.math.exp(T_x_y), axis=-2)
+        loss = tf.reduce_mean(T_xy, axis=-2) - tf.math.log(mean_exp_T_x_y)
+        loss = tf.squeeze(loss)  # Mutual Information
```

Comment:
to be safer in case some other dimension is 1: tf.squeeze(loss, axis=-1)

Reply:
I think tf.squeeze works fine here

Reply:
I know it works for your current setting. But it may not work as intended if batch_size is 1. In that case, the batch dimension will also be squeezed.

Reply:
use squeeze with axis = -1 now